File: C++/Problems/Problems/problem_274.cpp
#include "problem_274.h"

bool problem_274::test() {
    std::vector citations { 3, 0, 6, 1, 5 };

    int expected = 3;

    auto result = hIndex(citations);

    std::cout << result << std::endl;

    return result == expected;
}

int problem_274::hIndex(std::vector<int> &citations) {
    std::vector counts(1001, 0);
    for (auto c : citations)
        counts[c]++;
    int cur = 0;
    for (int i = 1000; i >= 0; --i)
    {
        cur += counts[i];
        if (cur >= i)
            return i;
    }

    return 0;
}

File: Python/Problems/Problems/problem274.py
import heapq
import math
import sys

from utils import print1DArray, print2DArray
from typing import List, Set, Optional
from collections import Counter

from .problem import Problem


class Problem274(Problem):
    def test(self) -> bool:
        citations = [3, 0, 6, 1, 5]

        expected = 3

        result = self.hIndex(citations)

        print(result)

        return result == expected

    def hIndex(self, citations: List[int]) -> int:
        counts = [0] * 1001
        for c in citations:
            counts[c] += 1

        cur = 0
        for i in range(1000, -1, -1):
            cur += counts[i]
            if cur >= i:
                return i

        return 0